README.md
# Proxywrap [![Build Status](https://travis-ci.org/findhit/proxywrap.svg?branch=master)](https://travis-ci.org/findhit/proxywrap)

Add PROXY v1 or v2 support to net, http, https, spdy and http2 servers. IPv4 and IPv6 protocols supported

## History

This module is a fork of original [proxywrap](https://github.com/cusspvz/proxywrap) by [Josh Dague](https://github.com/daguej) and [José Moreira
](https://github.com/cusspvz). Unfortunately, the project doesn't have recent changes. As so, we decided to contribute to it by forking it and make it better.

## What's the purpose of this module?

This module wraps node's various `Server` interfaces so that they are compatible with the [PROXY protocol](http://haproxy.1wt.eu/download/1.5/doc/proxy-protocol.txt).  It automatically parses the PROXY headers and resets `socket.remoteAddress` and `socket.remotePort` so that they have the correct values.

This module is especially useful if you need to get the client IP address when you're behind an AWS ELB in TCP mode.

In HTTP or HTTPS mode (aka SSL termination at ELB), the ELB inserts `X-Forwarded-For` headers for you.  However, in TCP mode, the ELB can't understand the underlying protocol, so you lose the client's IP address.  With the PROXY protocol and this module, you're able to retain the client IP address with any protocol.

In order for this module to work with ELB, you must [enable the PROXY protocol on your ELB](http://docs.aws.amazon.com/ElasticLoadBalancing/latest/DeveloperGuide/enable-proxy-protocol.html) (or whatever proxy your app is behind).

## Compability

This module is only compatible with **LTS** and **latest stable** versions of [node](https://github.com/nodejs/node).

## Installing

```bash
npm install --save @short.io/proxywrap
```

## Usage

**proxywrap** is a drop-in replacement.  Here's a simple Express app:

```js
var http = require( 'http' )
var proxiedHttp = require( 'findhit-proxywrap' ).proxy( http )
var express = require( 'express' )
var app = express()

// instead of http.createServer(app)
var srv = proxiedHttp.createServer( app ).listen( 80 )

app.get( '/', ( req, res ) => {
    res.send( 'IP = ' + req.connection.remoteAddress + ':' + req.connection.remotePort )
})
```

The magic happens in the `proxywrap.proxy()` call.  It wraps the module's `Server` constructor and handles a bunch of messy details for you.

You can do the same with `net` (raw TCP streams), `https`, and `spdy`.  It will probably work with other modules that follow the same pattern, but none have been tested.

*Note*: If you're wrapping [node-spdy](https://github.com/indutny/node-spdy), its exports are a little strange:

    var proxiedSpdy = require('proxywrap').proxy(require('spdy').server);

This also adds to all your sockets the properties:
* `socket.clientAddress` - The IP Address that connected to your PROXY.
* `socket.clientPort` - The Port used by who connected to your PROXY.
* `socket.proxyAddress` - The IP Address exposed on Client <-> Proxy side.
* `socket.proxyPort` - The Port exposed on Client <-> Proxy side. Usefull for detecting SSL on AWS ELB.
* `socket.remoteAddress` [optional] - Same as `socket.clientAddress`, used for compability proposes.
* `socket.remotePort` [optional] - Same as `socket.clientPort`, used for compability proposes.

**Warning:** By default, *all* traffic to your proxied server MUST use the PROXY protocol.  If the first five bytes received aren't `PROXY`, the connection will be dropped.  Obviously, the node server accepting PROXY connections should not be exposed directly to the internet; only the proxy (whether ELB, HAProxy, or something else) should be able to connect to node.

## API


### `proxy(Server[, options])`

Wraps something that inherits from the `net` module, exposing a `Server` and `createServer`.  Returns the same module patched to support the PROXY protocol.

Options:

- `strict` (default `true`): Incoming connections MUST use the PROXY protocol.  If the first five bytes received aren't `PROXY`, the connection will be dropped.  Disabling this option will allow connections that don't use the PROXY protocol (so long as the first bytes sent aren't `PROXY`).  Disabling this option poses a security risk; it should be enabled in production.

- `ignoreStrictExceptions` (default `false`): `strict` shutdowns your process with an error attached, meaning that if it isn't being caught on socket's `error` event, node will terminate process with an `uncaughtException`. This option tells `strict` methods to destroy sockets without providing the exception, so `node` ignores it. See [#11](https://github.com/findhit/proxywrap/issues/11) for more info.

- `overrideRemote` (default `true`): **findhit-proxywrap** overrides `socket.remoteAddress` and `socket.remotePort` for compability proposes. If you set this as `false`, your `socket.remoteAddress` and `socket.remotePort` will have the Address and Port of your **load-balancer** or whatever you are using behind your app. You can also access client's Address and Port by using `socket.clientAddress` and `socket.clientPort`.

## Contribute

Do you have any idea to improve this module?
Feel free to open an [Issue](https://github.com/findhit/proxywrap/issues/new) or a [Pull Request](https://github.com/findhit/proxywrap/pulls).

### People who have contributed so far

* [@cusspvz](https://github.com/cusspvz)
* [@daguej](https://github.com/daguej)
* [@revington](https://github.com/revington)
* [@mvalo](https://github.com/mvalo)
* [@sandfox](https://github.com/sandfox)
* [@xaka](https://github.com/xaka)

## Thanks

Thanks to all contibuters and special thanks to [Josh Dague](https://github.com/daguej) for creating original [proxywrap](https://github.com/daguej/node-proxywrap).


CHANGELOG.md
CHANGELOG
=========

# Upcoming

# 1.0.0-alpha.2
- Added http2 support


# 1.0.0-alpha.1
- Added support to proxy protocol v2
- Parsing proxy protocol via @balena/proxy-protocol-parser module instead of regex
- IPv6 support
- Updated code for latest ES features
- Fixed a bug when server.setTimeout did not trigger timeout event
- Removed dependency on object-extends and findhit-util

# 0.3.12
- Fixed HTTPS connection
- Updated spdy module version

# 0.3.11
- Fixed an issue where some connections were remaining opened. Thanks @revington [#15](https://github.com/findhit/proxywrap/issues/15)
- Added jshint to test procedure

# 0.3.10
- Added `.header` exposure on `error`
- Fixed TCP port on pr [#11](https://github.com/findhit/proxywrap/pull/11)

# 0.3.9
- Added an option to ignore strict generated exceptions while destroying socket,
  called `ignoreStrictExceptions`. Defaults to `false`. Reason on [#11](https://github.com/findhit/proxywrap/issues/11)

# 0.3.7
- Fixed destructed problem when on non-strict #7
- Updated dependent modules

# 0.3.6
- Fixed npm problem

# 0.3.5
- Implemented a better IPv6 detection approach
- Moved bluebird to dev dependencies #8
- Updated findhit-util lib to minor 2 updates

# 0.3.4
- Added RegExp Protocol detection

# 0.2.0 (2013-10-10)
- Added `options` parameter, ability to disable strict protocol checks.  Thanks to kylegetson.

# 0.1.2 (2013-08-01)
- Removed a `console.log` call that was accidentally left in.

# 0.1.1 (2013-07-31)
- Initial release.


test/tests.util.js
const ProxyWrap = require('../')
const fs = require('fs')
const proxyProtocol = require('@balena/proxy-protocol-parser')

function isSecureProtocol(protocol) {
  return protocol === 'https' || protocol == 'spdy' || protocol == 'http2';
}

const protocols = {
  net: require('net'),
  http: require('http'),
  https: require('https'),
  spdy: require('spdy'),
  http2: require('http2'),
}

const secureOptions = {
  key: fs.readFileSync('test/fixtures/key.pem'),
  cert: fs.readFileSync('test/fixtures/cert.pem')
}

const Chai = require('chai')
const expect = Chai.expect

module.defaults = {
  fakeConnect: {
    protocol: 'TCP4',
    autoCloseSocket: true,
    testAttributes: true,
    clientAddress: '10.10.10.1',
    proxyAddress: '10.10.10.254',
    remoteAddress: '10.10.10.1',
    clientPort: 12456,
    proxyPort: 80,
    headerJoinCRLF: true
  }
}

module.exports = {
  createServer: function(p, options) {
    const pc = protocols[p]
    const proxy = ProxyWrap.proxy(pc, options)
    const server = proxy.createServer(isSecureProtocol(p) ? secureOptions : {})
    const port = Math.floor(Math.random() * 5000 + 20000) // To be sure that the port is not beeing used on test side
    const host = '127.0.0.1'

    server._protocol = p
    server._protocolConstructor = pc
    server.host = host
    server.port = port

    // Start server on localhost/random-port
    server.listen(port, host)

    // Returns server
    return server
  },

  fakeConnect: function(server, options) {
    const p = server._protocol;

    // Prepare options
    options = {
      ...module.defaults.fakeConnect,
      ...options,
    };
    let header;
    if (!options.header) {
      header = proxyProtocol[`v${options.protocolVersion || 1}_encode`]({
        remoteAddress: options.clientAddress,
        remotePort: options.clientPort,
        localAddress: options.proxyAddress,
        localPort: options.proxyPort,  
      });
    } else {
      header = Buffer.from(options.header)
    }

    const body = Buffer.from(['GET /something/cool HTTP/1.1', 'Host: www.findhit.com'].join('\n'))

    return (new Promise(function ( fulfill, reject ) {
      if ( typeof server.listening == 'boolean' ) {
        if ( server.listening ) {
          fulfill()
        } else {
          server.once('listening', fulfill)
          server.once('error', reject)
        }
      } else {
        fulfill()
      }
    }))
    .then(function () {
      return new Promise(function(fulfill, reject) {
        const client = new protocols.net.Socket(),
          host = server.host,
          port = server.port
        const value = [undefined, client]

        server.once('connection', function(socket) {
          socket.on('error', function(err) {
            reject(err)
          })
        })

        server.once('proxiedConnection', function(socket) {
          value[0] = socket

          socket.on('error', function(err) {
            reject(err)
          })

          if (options.testAttributes && !options.header) {
            try {
              expect(socket.clientAddress).to.be.equal(options.clientAddress, 'Client address does not match')
              expect(socket.proxyAddress).to.be.equal(options.proxyAddress, 'Proxy address does not match')
              expect(socket.clientPort).to.be.equal(options.clientPort, 'Client port does not match')
              expect(socket.proxyPort).to.be.equal(options.proxyPort, 'Proxy port does not match')
              if (server.proxyOptions.overrideRemote) {
                expect(socket.remoteAddress).to.be.equal(options.clientAddress, 'Remote address does not match')
                expect(socket.remotePort).to.be.equal(options.clientPort, 'Remote port does not match')
              }
            } catch (err) {
              reject(err)
            }
          }

          if (options.autoCloseSocket && !isSecureProtocol(p)) {
            socket.end()
          } else {
            fulfill(value)
          }
        })

        client.once('connect', function() {
          // Send header and body
          client.write(Buffer.concat([header, body]))
        })

        client.connect(port, host)

        if (options.autoCloseSocket) {
          client.once('end', function() {
            fulfill(value)
          })
        }
      })
    })
  }
}


test/proxy.protocol.test.js
const tUtil = require('./tests.util'),
  net = require('net'),
  chai = require('chai'),
  expect = chai.expect;
const proxyProtocol = require('@balena/proxy-protocol-parser');

for (const serverType of ['net', 'http', 'https', 'spdy', 'http2']) {
  for (const protocolVersion of [1, 2]) {
    const proxyProtocolEncode = proxyProtocol[`v${protocolVersion}_encode`];
    describe(`PROXY Protocol v${protocolVersion} server: ${serverType}`, function () {

      const server = tUtil.createServer(serverType, { strict: true, protocolVersion })
      it('Check socket is established correctly', function () {
        return tUtil.fakeConnect(server, { protocolVersion })
      })

      it('Check with another socket parameters', function () {
        return tUtil.fakeConnect(server, {
          headers: proxyProtocolEncode({
            localAddress: '192.168.0.1',
            localPort: 3350,
            remoteAddress: '192.168.0.254',
            remotePort: 443
          })
        })
      })

      it(`Check with another socket parameters as a string in v${protocolVersion} format`, function () {
        return tUtil.fakeConnect(server, {
          header: proxyProtocolEncode({
            remoteFamily: 'IPv4',
            remoteAddress: '192.168.0.254',
            localAddress: '192.168.0.1',
            localPort: 443,
            remotePort: 3350,
          }),
        })
      })

      it('Check with IPv6 IP', function () {
        return tUtil.fakeConnect(server, {
          headers: proxyProtocolEncode({
            family: 'IPv6',
            localAddress: 'fe80::a00:27ff:fe9f:4016',
            localPort: 3350,
            remoteAddress: 'fe80::a089:a3ff:fe15:e992',
            remotePort: 443
          })
        })
      })

      describe('Should detect a malformed PROXY headers', function () {
        it("Header without IP's", function () {
          return tUtil
            .fakeConnect(server, {
              header: 'PROXY HACK ATTEMPT\r\n'
            })
            .then(
              function () {
                throw new Error("It shouldn't get fulfilled")
              },
              function (err) {
                expect(err.message).to.be.equal('PROXY protocol malformed header')
              }
            )
        })

        if (serverType === 'net') {
          it('non-proxy connection when in non-strict mode should not be destroyed #7', function () {
            return tUtil.fakeConnect(tUtil.createServer(serverType, { strict: false }), {
              header: 'TELNET BABY'
            })
          })
        }

        it('Restore emitted events after socket.destroy #5', function () {
          return tUtil
            .fakeConnect(server, {
              header: 'PRO',
              autoCloseSocket: false,
              testAttributes: false
            })
            .then(
              function () {
                throw new Error("It shouldn't get fulfilled")
              },
              function (err) {
                expect(err.message).to.be.equal('non-PROXY protocol connection')
              }
            )
        })

        it('should drop connection gracefully when non-proxy connection is gathered when `ignoreStrictExceptions` is active. #11', function (
          cb
        ) {
          const server = tUtil.createServer(serverType, {
            strict: true,
            ignoreStrictExceptions: true
          })

          server.once('listening', function () {
            const client = new net.Socket()

            client.on('end', cb)

            client.once('connect', function () {
              // Send header and body
              client.write('GET / HTTP/1.0\n\n')
            })

            client.connect(server.port, server.host)
          })

        })
        if (serverType !== 'net') {
          it('proxy socket timeout should work', (done) => {
            const server = tUtil.createServer(serverType, {
              strict: false,
              ignoreStrictExceptions: true
            });
            server.setTimeout(500);
            expect(server.timeout).to.equal(500);

            server.once('listening', function () {
              const client = new net.Socket()

              client.on('end', done)

              client.once('connect', function () {
                client.write('GET /')
              })

              client.connect(server.port, server.host)
            })
          })

          /*
          
          it('socket timeout should work', (done) => {
            const server = require(serverType).createServer();
            const port = Math.floor(Math.random() * 5000 + 20000) ;
            server.listen(port);
            server.setTimeout(500);
            expect(server.timeout).to.equal(500);

            server.once('listening', function () {
              const client = new net.Socket()

              client.on('end', done)

              client.once('connect', function () {
                client.write('GET /')
              })

              client.connect(port, server.host)
            })
          })
          */
        }
      })
    });
  }
}


test/proxy.protocol.v1.test.js
var Util = require('findhit-util'),
  tUtil = require('./tests.util'),
  net = require('net'),
  chai = require('chai'),
  expect = chai.expect;
const proxyProtocol = require('proxy-protocol-v2')

describe('PROXY Protocol v1', function() {
  describe('net', function() {
    var server = tUtil.createServer('net', { strict: true })
    v1tests('net', server)
  })

  describe('http', function() {
    var server = tUtil.createServer('http', { strict: true })
    v1tests('http', server)
  })

  describe('https', function() {
    var server = tUtil.createServer('https', { strict: true })
    v1tests('https', server)
  })

  describe('spdy', function() {
    var server = tUtil.createServer('spdy', { strict: true })
    v1tests('spdy', server)
  })
})

function v1tests(proto, server) {
  it('Check socket is stablished correctly', function() {
    return tUtil.fakeConnect(server)
  })

  it('Check with another socket parameters', function() {
    return tUtil.fakeConnect(server, {
      clientAddress: '192.168.0.1',
      clientPort: 3350,
      proxyAddress: '192.168.0.254',
      proxyPort: 443
    })
  })

  it('Check with another socket parameters as a string in v1 format', function() {
    return tUtil.fakeConnect(server, {
      header: proxyProtocol.v1_encode({
        remoteFamily: 'IPv4',
        remoteAddress: '192.168.0.254',
        localAddress: '192.168.0.1',
        localPort: 443,
        remotePort: 3350,
      })
    })
  })

  describe('Should detect a malformed PROXY headers', function() {
    it("Header without IP's", function() {
      return tUtil
        .fakeConnect(server, {
          header: 'PROXY HACK ATTEMPT'
        })
        .then(
          function() {
            throw new Error("It shouldn't get fulfilled")
          },
          function(err) {
            expect(err.message).to.be.equal('PROXY protocol malformed header')
          }
        )
    })

    if (proto === 'net') {
      it('non-proxy connection when in non-strict mode should not be destroyed #7', function() {
        return tUtil.fakeConnect(tUtil.createServer(proto, { strict: false }), {
          header: 'TELNET BABY'
        })
      })
    }

    it('Restore emitted events after socket.destroy #5', function() {
      return tUtil
        .fakeConnect(server, {
          header: 'PRO',
          autoCloseSocket: false,
          testAttributes: false
        })
        .then(
          function() {
            throw new Error("It shouldn't get fulfilled")
          },
          function(err) {
            expect(err.message).to.be.equal('non-PROXY protocol connection')
          }
        )
    })

    it('should drop connection gracefully when non-proxy connection is gathered when `ignoreStrictExceptions` is active. #11', function(
      cb
    ) {
      var server = tUtil.createServer(proto, {
        strict: true,
        ignoreStrictExceptions: true
      })

      server.once('listening', function () {
        var client = new net.Socket()

        client.on('end', cb)

        client.once('connect', function() {
          // Send header and body
          client.write('GET / HTTP/1.0\n\n')
        })

        client.connect(server.port, server.host)
      })

    })
  })
}


lib/proxy-protocol.regexp.js
/*
[source](http://www.haproxy.org/download/1.5/doc/proxy-protocol.txt)

2.1

This is the format specified in version 1 of the protocol. It consists in one
line of ASCII text matching exactly the following block, sent immediately and
at once upon the connection establishment and prepended before any data flowing
from the sender to the receiver :

- a string identifying the protocol : "PROXY" ( \x50 \x52 \x4F \x58 \x59 )
Seeing this string indicates that this is version 1 of the protocol.

- exactly one space : " " ( \x20 )

- a string indicating the proxied INET protocol and family. As of version 1,
only "TCP4" ( \x54 \x43 \x50 \x34 ) for TCP over IPv4, and "TCP6"
( \x54 \x43 \x50 \x36 ) for TCP over IPv6 are allowed. Other, unsupported,
or unknown protocols must be reported with the name "UNKNOWN" ( \x55 \x4E
\x4B \x4E \x4F \x57 \x4E ). For "UNKNOWN", the rest of the line before the
CRLF may be omitted by the sender, and the receiver must ignore anything
presented before the CRLF is found. Note that an earlier version of this
specification suggested to use this when sending health checks, but this
causes issues with servers that reject the "UNKNOWN" keyword. Thus is it
now recommended not to send "UNKNOWN" when the connection is expected to
be accepted, but only when it is not possible to correctly fill the PROXY
line.

- exactly one space : " " ( \x20 )

- the layer 3 source address in its canonical format. IPv4 addresses must be
indicated as a series of exactly 4 integers in the range [0..255] inclusive
written in decimal representation separated by exactly one dot between each
other. Heading zeroes are not permitted in front of numbers in order to
avoid any possible confusion with octal numbers. IPv6 addresses must be
indicated as series of 4 hexadecimal digits (upper or lower case) delimited
by colons between each other, with the acceptance of one double colon
sequence to replace the largest acceptable range of consecutive zeroes. The
total number of decoded bits must exactly be 128. The advertised protocol
family dictates what format to use.

- exactly one space : " " ( \x20 )

- the layer 3 destination address in its canonical format. It is the same
format as the layer 3 source address and matches the same family.

- exactly one space : " " ( \x20 )

- the TCP source port represented as a decimal integer in the range
[0..65535] inclusive. Heading zeroes are not permitted in front of numbers
in order to avoid any possible confusion with octal numbers.

- exactly one space : " " ( \x20 )

- the TCP destination port represented as a decimal integer in the range
[0..65535] inclusive. Heading zeroes are not permitted in front of numbers
in order to avoid any possible confusion with octal numbers.

- the CRLF sequence ( \x0D \x0A )


The maximum line lengths the receiver must support including the CRLF are :
- TCP/IPv4 :
"PROXY TCP4 255.255.255.255 255.255.255.255 65535 65535\r\n"
=> 5 + 1 + 4 + 1 + 15 + 1 + 15 + 1 + 5 + 1 + 5 + 2 = 56 chars

- TCP/IPv6 :
"PROXY TCP6 ffff:f...f:ffff ffff:f...f:ffff 65535 65535\r\n"
=> 5 + 1 + 4 + 1 + 39 + 1 + 39 + 1 + 5 + 1 + 5 + 2 = 104 chars

- unknown connection (short form) :
"PROXY UNKNOWN\r\n"
=> 5 + 1 + 7 + 2 = 15 chars

- worst case (optional fields set to 0xff) :
"PROXY UNKNOWN ffff:f...f:ffff ffff:f...f:ffff 65535 65535\r\n"
=> 5 + 1 + 7 + 1 + 39 + 1 + 39 + 1 + 5 + 1 + 5 + 2 = 107 chars

*/

var IPv4 =
  '(((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?).){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?))',
  IPv6 = '([0-9a-fA-F]{0,4}:){2,7}(:|[0-9a-fA-F]{1,4})',
  Port =
    '([1-9][0-9]{0,3}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5])' // 1..65535

var ProxyProtocolRegexp = new RegExp(
  '^PROXY (' +
  'UNKNOWN' +
  '.{0,94}' +
  '|' /* 107 - 13 */ +
    ['TCP4', IPv4, IPv4, Port, Port].join(' ') +
    '|' +
    ['TCP6', IPv6, IPv6, Port, Port].join(' ') +
    ')$'
)

// Export it
module.exports = ProxyProtocolRegexp


lib/proxywrap.js
'use strict'
/*
 * node-proxywrap
 *
 * Copyright (c) 2013, Josh Dague
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

const proxyProtocol = require('@balena/proxy-protocol-parser');
const proxyProtocolCommon = require('@balena/proxy-protocol-parser/lib/v2_common');

exports.defaults = {
  strict: true,
  ignoreStrictExceptions: false,
  overrideRemote: true,
};

const v1Header = 'PROXY';
const v2Header = proxyProtocolCommon.sigBytes;

const proxyProtocolFields = [
  'remoteAddress',
  'remotePort',
  'clientAddress',
  'clientPort',
  'proxyAddress',
  'proxyPort',
];

const isHeaderCompleted = (buf) => {
  if (buf.slice(0, 5).equals(Buffer.from('PROXY'))) {
    const endOfBufferIndex = buf.indexOf('\r');
    if (endOfBufferIndex >= 0) {
      const proxyInfo = proxyProtocol.v1_decode(buf.slice(0, endOfBufferIndex));

      return [true, proxyInfo, buf.slice(endOfBufferIndex + 2)];
    }
    return [false, null, buf.slice(endOfBufferIndex + 2)];
  }
  if (buf.slice(0, v2Header.length).equals(v2Header)) {
    const addrLength = buf[15] + buf[14] * 256;
    const proxyInfo = proxyProtocol.v2_decode(buf.slice(0, 16 + addrLength));
    return [true, proxyInfo, buf.slice(16 + addrLength)]
  }
  return [false, null, buf]
}

function createTLSSocketPropertyGetter(tlsSocket, propertyName) {
  return function () {
    return tlsSocket._parent[propertyName];
  };
}

function defineProperty(target, propertyName, getter) {
  Object.defineProperty(target, propertyName, {
    enumerable: false,
    configurable: true,
    get: getter,
  });
}

function defineSocketProperties(socket, proxyInfo, overrideRemote) {
  const socketParams = {
    clientAddress: proxyInfo.remoteAddress,
    proxyAddress: proxyInfo.localAddress,
    clientPort: proxyInfo.remotePort,
    proxyPort: proxyInfo.localPort,
  };
  for (const [propertyName, propertyValue] of Object.entries(socketParams)) {
    defineProperty(socket, propertyName, () => propertyValue);
  }
  if (overrideRemote) {
    defineProperty(socket, 'remoteAddress', () => socketParams.clientAddress);
    defineProperty(socket, 'remotePort', () => socketParams.clientPort);
  }
}

// Wraps the given module (ie, http, https, net, tls, etc) interface so that
// `socket.remoteAddress` and `remotePort` work correctly when used with the
// PROXY protocol (http://haproxy.1wt.eu/download/1.5/doc/proxy-protocol.txt)
// strict option drops requests without proxy headers, enabled by default to match previous behavior, disable to allow both proxied and non-proxied requests
exports.proxy = function (iface, options) {

  const replaceListeners = (server) => {
    // remove the connection listener attached by iface.Server and replace it with our own.
    const connectionListeners = server.listeners('connection');
    server.removeAllListeners('connection');
    server.addListener('connection', connectionListener);
  
    // add the old connection listeners to a custom event, which we'll fire after processing the PROXY header
    for (const connectionListener of connectionListeners) {
      server.addListener('proxiedConnection', connectionListener);
    }
  
    // changing secure connection listeners to set remoteAddress property on socket
    const secureConnectionListeners = server.listeners('secureConnection');
    server.removeAllListeners('secureConnection');
  
    for (const secureConnectionListener of secureConnectionListeners) {
      server.addListener(
        'secureConnection',
        createSecureConnectionListener(server, secureConnectionListener),
      );
    }
  }

  const exports = {};

  options = {
    ...module.exports.defaults,
    ...options
  };

  // copy iface's exports to myself
  for (const k in iface) exports[k] = iface[k];
  ['createServer', 'createSecureServer'].filter((method) => iface[method]).forEach((method) => {
    exports[method] = function (...args) {
      const server = iface[method](...args);
      server.proxyOptions = options;
      replaceListeners(server);
      return server;
    };
  })


  exports.options = options;

  function connectionListener(socket) {
    const self = this;
    const realEmit = socket.emit;
    let history = [];
    let protocolError = false;
    if (this.timeout && socket.timeout === undefined) {
      socket.setTimeout(this.timeout, socket.end);
    }
    // override the socket's event emitter so we can process data (and discard the PROXY protocol header) before the underlying Server gets it
    socket.emit = (function () {
      let isReadable;
      return (event, ...rest) => {
        history.push([event, ...rest]);

        if (event === 'readable') {
          isReadable = true;
          return onReadable();
        }
        // Only needed for node.js 0.10
        if (event === 'end' && !isReadable) {
          self.emit('proxiedConnection', socket);
          restore();
        }
        if (event === 'timeout') {
          realEmit.apply(socket, [event]);
        }
      };
    }());

    function restore() {
      if (socket.emit === realEmit) return;

      // if (legacy) socket.removeListener('data', ondata);
      // restore normal socket functionality, and fire any events that were emitted while we had control of emit()
      socket.emit = realEmit;
      for (let i = 0; i < history.length; i++) {
        realEmit.apply(socket, history[i]);
        if (history[i][0] == 'end' && socket.onend) socket.onend();
      }
      history = null;
    }

    function destroy(error, wasStrict) {
      error = error || undefined;

      if (!(error instanceof Error)) {
        error = new Error(error);
      }

      // Set header on error
      error.header = buf.toString('ascii');

      protocolError = true;

      socket.destroy(
        wasStrict
          ? (!options.ignoreStrictExceptions && error) || undefined
          : error,
      );

      restore();
    }

    socket.on('readable', onReadable);

    let buf = Buffer.alloc(0);

    function onReadable() {
      let chunk;
      chunk = socket.read();

      if (chunk === null && buf.length === 0) {
        // unshifting will fire the readable event
        socket.emit = realEmit;
        self.emit('proxiedConnection', socket);
        return;
      }

      while (chunk !== null) {
        buf = Buffer.concat([buf, chunk]);
        // if the first 5 bytes aren't PROXY, something's not right.
        if (
          buf.length >= Math.max(v1Header.length, v2Header.length) &&
          (
            !buf.slice(0, v1Header.length).equals(Buffer.from(v1Header)) &&
            !buf.slice(0, v2Header.length).equals(Buffer.from(v2Header))
          )
        ) {
          protocolError = true;
          if (options.strict) {
            return destroy('non-PROXY protocol connection', true);
          }
        }
        const [headerCompleted, proxyInfo, bufferRest] = isHeaderCompleted(buf);
        if (headerCompleted || protocolError) {
          socket.removeListener('readable', onReadable);

          if (options.strict) {
            if (!proxyInfo || isNaN(proxyInfo.remotePort)) {
              return destroy('PROXY protocol malformed header', true);
            }
          }

          if (!protocolError) {
            defineSocketProperties(socket, proxyInfo, options.overrideRemote);
          }

          // unshifting will fire the readable event
          socket.emit = realEmit;
          socket.unshift(bufferRest);

          self.emit('proxiedConnection', socket);

          restore();

          if (socket.ondata) {
            const data = socket.read();

            if (data) {
              socket.ondata(data, 0, data.length);
            }
          }

          return;
        } if (buf.length > 107) {
          return destroy('PROXY header too long', false);
        }

        chunk = socket.read();
      }
    }
  }

  function createSecureConnectionListener(context, listener) {
    return function (socket) {
      const properties = proxyProtocolFields;
      defineTLSSocketProperties(socket, properties);
      listener.call(context, socket);
    };
  }

  function defineTLSSocketProperties(tlsSocket, properties) {
    for (const propertyName of properties) {
      const getter = createTLSSocketPropertyGetter(tlsSocket, propertyName);
      defineProperty(tlsSocket, propertyName, getter);
    }
  }

  return exports;
};
